### DISCUSSION TITLE
Is there a way I can create a new column using pandas 'pd.cut' and 'pd.qcut' functions in visidata on a specific column?

### DISCUSSION THREAD

**Discussion by oostopitre** [saulpw/visidata#2613] [Q&A]
Is there a way I can create a new column in visidata using pandas 'pd.cut' and 'pd.qcut' functions on an existing numeric column?

I tried importing pandas in my .visidatarc but I am not able to figure out how to use these specific pandas functions on a column. I tried searching similar threads but wasn't able to figure out quite how to accomplish this. I think this might be because to use these functions, it needs info on the values of the whole column. 

Any tips or help is super appreciated. Love visidata!

**Comment by midichef** [comment]
Hi, this should work:

In your .visidatarc, put `import pandas as pd`
Make a Python list holding the current column's values:
`Space` `exec-python` `vals = [vd.sheet.cursorCol.getTypedValue(r) for r in sheet.rows]`
Process it with Pandas `cut()`:
`Space` `exec-python` `cutvals = pd.cut(vals, bins=[0,3,6,9], labels=['one','two','three'])`
Then make a new column whose rows are the elements of `cutvals`:
`Space` `addcol-iter` `cutvals`

If you get an error like `TypeError: '<' not supported between instances of 'int' and 'str'`, the problem may be that your starting column was a string. Make sure it has a numeric type, by pressing `#` or `%` or `$`, then remake `vals` using `exec-python`.

If you want to run this command often, [make a macro](https://www.visidata.org/docs/macros/), and generalize the cut command to refer to your own custom variables, e.g. `cutbins` and `cutlabels`:
`Space` `exec-python` `cutvals = pd.cut(vals, bins=cutbins, labels=cutlabels)`

It may be possible in the future to make a new column more simply, with some changes to aggregators that are still being considered.

**Reply by oostopitre** [reply]
Thank you for the suggestion. Will try this out. This will give me atleast a temporary way to do this.

**Comment by midichef** [comment]
Here is a custom command `addcol-pandas-cut`, to run `pd.cut` on the current column. It prompts you to enter bins and labels manually. And it would be easy to modify it to use predefined bins and labels.
```
@Sheet.api
def addcol_pandas_cut(sheet, col, cut_bins=None, cut_labels=None):
    import pandas as pd
    vals = [col.getTypedValue(r) for r in sheet.rows]
    if not cut_bins:
        cut_bins = eval(sheet.inputExpr('bins expr: '))
    if not cut_labels:
        cut_labels = eval(sheet.inputExpr(f'{(len(cut_bins) - 1)} labels expr: '))
    cutvals = pd.cut(vals, bins=cut_bins, labels=cut_labels)
    c = SettableColumn()
    sheet.addColumnAtCursor(c)
    c.setValues(sheet.rows, *cutvals)
Sheet.addCommand('', 'addcol-pandas-cut', 'sheet.addcol_pandas_cut(cursorCol)', 'custom;  apply pd.cut to current column')
```
